### PR TITLE
Allman style formatting (JavaScript files)

### LINKED ISSUE #1792: Allman style formatting
# Description
**Allman style formatting required as below:**
1) start and end braces on separate lines
2) start and end square brackets on separate lines
3) array items on new line

### PR DESCRIPTION
**Pull request by UAK-35**
# Description
- [ ] Source branch in your fork has meaningful name (not `master`)

This is an attempt to add Allman style formatting to beautifier for JavaScript files

**New feature performs following formatting:**
1. Start and end braces on separate lines
2. Start and end square brackets on separate lines
3. Array items on new line

Fixes Issue: 
Closes #1792

[Maintainability](https://codeclimate.com/github/UAK-35/js-beautify/maintainability)

New option for brace style added is : **expand-all**

# Before Merge Checklist 
These items can be completed after PR is created.

(Check any items that are not applicable (NA) for this PR)

- [ ] JavaScript implementation
- [ ] Python implementation (NA if HTML beautifier)
- [x] Added Tests to data file(s)
- [ ] Added command-line option(s) (NA if
- [ ] README.md documents new feature/option(s)

### CODE REVIEW

**Comment by UAK-35** [comment]
### **CONSIDER THE FIRST 3 COMMITS ONLY AS THEY CONTAIN THE ACTUAL MODIFICATIONS TO CODE**

Hashes of important commits are:

1. Added option to use Allman style (535b2ee82ae6b00c608370185b95208eb1f6ea31)
2. Added option to use Allman style for python - Ensuring Feature Parity (74fa063ff06e1540ff67dacd4920019d2280fae6)
3. Documentation/Help added for new option to use Allman style (8cbe03a15f73450d2b3ead54127f6585ec8715f4)

**After the 3rd commit (8cbe03a15f73450d2b3ead54127f6585ec8715f4), modifications are complete**

**BELOW IS THE FINAL COMMIT CONTAINING ALL CODE MODIFICATIONS AS PER GUIDELINES:**
_Documentation/Help added for new option to use Allman style (8cbe03a15f73450d2b3ead54127f6585ec8715f4)_


_

> **OTHER COMMITS ARE JUST FOR PASSING AZURE PIPLINES RUNS AND CODE CLIMATE TEST PASSING**

_

File: python/jsbeautifier/javascript/beautifier.py [hunk 1]
```diff
@@ -583,6 +599,11 @@ def handle_end_expr(self, current_token):
                 current_token, current_token.text == ']' and self.is_array(
                     self._flags.mode) and not self._options.keep_array_indentation)
 
+        if self._options.brace_style == "expand-all" and \
+                current_token.text == ']' and \
+                self.is_array(self._flags.mode):
+            self.print_newline()
+
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:606]: New feature to put ending square bracket on its own separate line (Newline before)

File: python/jsbeautifier/javascript/beautifier.py [hunk 2]
```diff
@@ -561,6 +574,9 @@ def handle_start_expr(self, current_token):
                 current_token, current_token.newlines)
 
         self.print_token(current_token)
+        if current_token.text == '[' and \
+                self._options.brace_style == "expand-all":
+            self.print_newline()
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:579]: New feature to put starting square bracket on its own separate line (Newline afterwards)

File: python/jsbeautifier/javascript/beautifier.py [hunk 3]
```diff
@@ -670,8 +691,8 @@ def handle_start_block(self, current_token):
     check_token.type != TOKEN.EOF and not (
          check_token.type == TOKEN.END_BLOCK and check_token.opened == current_token))
 
-        if (self._options.brace_style == 'expand' or (self._options.brace_style ==
-                                                  'none' and current_token.newlines)) and not self._flags.inline_frame:
+        if (self._options.brace_style == 'expand' or self._options.brace_style == 'expand-all' or
+                (self._options.brace_style == 'none' and current_token.newlines)) and not self._flags.inline_frame:
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:695]: New feature to have same behavior as that of brace style option with value expand.

File: python/jsbeautifier/javascript/beautifier.py [hunk 4]
```diff
@@ -720,7 +741,7 @@ def handle_end_block(self, current_token):
         # try inline_frame (only set if opt.braces-preserve-inline) first
         if self._flags.inline_frame and not empty_braces:
             self._output.space_before_token = True
-        elif self._options.brace_style == 'expand':
+        elif self._options.brace_style == 'expand' or self._options.brace_style == 'expand-all':
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:744]: New feature to have same behavior as that of brace style option with value expand.

File: python/jsbeautifier/javascript/beautifier.py [hunk 5]
```diff
@@ -868,9 +890,8 @@ def handle_word(self, current_token):
         elif self._flags.last_token.type == TOKEN.STRING:
             prefix = 'NEWLINE'
         elif self._flags.last_token.type == TOKEN.RESERVED or self._flags.last_token.type == TOKEN.WORD or \
-            (self._flags.last_token.text == '*' and (
-                self._last_last_text in ['function', 'yield'] or
-                (self._flags.mode == MODE.ObjectLiteral and self._last_last_text in ['{', ',']))):
+                (self._flags.last_token.text == '*' and (
+                        self._last_last_text in ['function', 'yield'] or is_literal_sbrk_cma)):
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:894]: UNNECESSARY

File: python/jsbeautifier/javascript/beautifier.py [hunk 6]
```diff
@@ -888,11 +909,10 @@ def handle_word(self, current_token):
...
 
         if reserved_array(current_token, ['else', 'catch', 'finally']):
-            if ((not (self._flags.last_token.type == TOKEN.END_BLOCK and self._previous_flags.mode == MODE.BlockStatement))
-                or self._options.brace_style == 'expand'
-                or self._options.brace_style == 'end-expand'
-                or (self._options.brace_style == 'none' and current_token.newlines)) \
-               and not self._flags.inline_frame:
+            blk_stmt_ends = self._flags.last_token.type == TOKEN.END_BLOCK
+            blk_stmt_ends = blk_stmt_ends and self._previous_flags.mode == MODE.BlockStatement
+            if ((not blk_stmt_ends) or self._options.brace_style in ['expand', 'end-expand', 'expand-all']
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:914]: New feature to have same behavior as that of brace style option with value expand or end-expand.

File: python/jsbeautifier/javascript/beautifier.py [hunk 7]
```diff
@@ -888,11 +909,10 @@ def handle_word(self, current_token):
                 prefix = 'NEWLINE'
 
         if reserved_array(current_token, ['else', 'catch', 'finally']):
-            if ((not (self._flags.last_token.type == TOKEN.END_BLOCK and self._previous_flags.mode == MODE.BlockStatement))
-                or self._options.brace_style == 'expand'
-                or self._options.brace_style == 'end-expand'
-                or (self._options.brace_style == 'none' and current_token.newlines)) \
-               and not self._flags.inline_frame:
+            blk_stmt_ends = self._flags.last_token.type == TOKEN.END_BLOCK
+            blk_stmt_ends = blk_stmt_ends and self._previous_flags.mode == MODE.BlockStatement
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:913]: UNNECESSARY

File: python/jsbeautifier/javascript/beautifier.py [hunk 8]
```diff
@@ -888,11 +909,10 @@ def handle_word(self, current_token):
...
+            if ((not blk_stmt_ends) or self._options.brace_style in ['expand', 'end-expand', 'expand-all']
+                    or (self._options.brace_style == 'none' and current_token.newlines)) and not self._flags.inline_frame:
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:915]: UNNECESSARY

File: js/src/javascript/beautifier.js [hunk 9]
```diff
@@ -973,11 +968,9 @@ Beautifier.prototype.handle_word = function(current_token) {
   }
 
   if (reserved_array(current_token, ['else', 'catch', 'finally'])) {
-    if ((!(this._flags.last_token.type === TOKEN.END_BLOCK && this._previous_flags.mode === MODE.BlockStatement) ||
-        this._options.brace_style === "expand" ||
-        this._options.brace_style === "end-expand" ||
-        (this._options.brace_style === "none" && current_token.newlines)) &&
-      !this._flags.inline_frame) {
+    var blockStatementEnds = this._flags.last_token.type === TOKEN.END_BLOCK && this._previous_flags.mode === MODE.BlockStatement;
+    if ((!blockStatementEnds || in_array(this._options.brace_style, ["expand", 'end-expand', 'expand-all']) ||
```
Comment (UAK-35) [inline js/src/javascript/beautifier.js:972]: New feature to have same behavior as that of brace style option with value expand or end-expand.

File: js/src/javascript/beautifier.js [hunk 10]
```diff
@@ -973,11 +968,9 @@ Beautifier.prototype.handle_word = function(current_token) {
...
+        (this._options.brace_style === "none" && current_token.newlines)) && !this._flags.inline_frame) {
```
Comment (UAK-35) [inline js/src/javascript/beautifier.js:973]: UNNECESSARY

File: js/src/javascript/beautifier.js [hunk 11]
```diff
@@ -973,11 +968,9 @@ Beautifier.prototype.handle_word = function(current_token) {
   }
 
   if (reserved_array(current_token, ['else', 'catch', 'finally'])) {
-    if ((!(this._flags.last_token.type === TOKEN.END_BLOCK && this._previous_flags.mode === MODE.BlockStatement) ||
-        this._options.brace_style === "expand" ||
-        this._options.brace_style === "end-expand" ||
-        (this._options.brace_style === "none" && current_token.newlines)) &&
-      !this._flags.inline_frame) {
+    var blockStatementEnds = this._flags.last_token.type === TOKEN.END_BLOCK && this._previous_flags.mode === MODE.BlockStatement;
```
Comment (UAK-35) [inline js/src/javascript/beautifier.js:971]: UNNECESSARY

File: python/jsbeautifier/javascript/options.py [hunk 12]
```diff
@@ -54,7 +54,7 @@ def __init__(self, options=None):
         # preserve-inline in delimited string will trigger brace_preserve_inline, everything
         # else is considered a brace_style and the last one only will have an effect
 
-        brace_style_split = self._get_selection_list('brace_style', ['collapse', 'expand', 'end-expand', 'none', 'preserve-inline'])
+        brace_style_split = self._get_selection_list('brace_style', ['collapse', 'expand', 'end-expand', 'none', 'preserve-inline', 'expand-all'])
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/options.py:57]: New feature is here added to JS-Beautify tool.

File: setup.cfg [hunk 13]
```diff
@@ -0,0 +1,2 @@
+[pycodestyle]
+max-line-length = 5000
```
Comment (UAK-35) [inline setup.cfg:2]: CodeClimate checks (Python specific) do not pass without these changes so for other developers these are required.

File: tox.ini [hunk 14]
```diff
@@ -0,0 +1,2 @@
+[pycodestyle]
+max-line-length = 5000
```
Comment (UAK-35) [inline tox.ini:2]: CodeClimate checks (Python specific) do not pass without these changes so for other developers these are required.

File: tools/template/beautify.wrapper.js [hunk 15]
```diff
@@ -61,8 +61,8 @@
     space_after_anon_function (default false) - should the space before an anonymous function's parens be added, "function()" vs "function ()",
           NOTE: This option is overriden by jslint_happy (i.e. if jslint_happy is true, space_after_anon_function is true by design)
 
-    brace_style (default "collapse") - "collapse" | "expand" | "end-expand" | "none" | any of the former + ",preserve-inline"
-            put braces on the same line as control statements (default), or put braces on own line (Allman / ANSI style), or just put end braces on own line, or attempt to keep them where they are.
+    brace_style (default "collapse") - "collapse" | "expand" | "end-expand" | "expand-all" | "none" | any of the former + ",preserve-inline"
+            put braces on the same line as control statements (default), or put braces on own line (Allman / ANSI style), or just put end braces on own line, or put braces/brackets/array-items on own line or attempt to keep them where they are.
```
Comment (UAK-35) [inline tools/template/beautify.wrapper.js:65]: New feature command line help added

File: python/jsbeautifier/javascript/beautifier.py [hunk 16]
```diff
@@ -848,18 +869,19 @@ def handle_word(self, current_token):
...
-                if self._options.brace_style in ['expand', 'end-expand'] or (
-                        self._options.brace_style == 'none' and current_token.newlines):
-                    prefix = 'NEWLINE'
-                else:
-                    prefix = 'SPACE'
-                    self._output.space_before_token = True
+        is_literal_sbrk_cma = self._flags.mode == MODE.ObjectLiteral and self._last_last_text in ['{', ',']
+        token_reserved_word = reserved_array(current_token, ['else', 'catch', 'finally', 'from'])
+        no_brace_style_nl = self._options.brace_style == 'none' and current_token.newlines
+        brace_style_in_exp = self._options.brace_style in ['expand', 'end-expand', 'expand-all']
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:875]: New feature to have same behavior as that of brace style option with value expand or end-expand.

File: python/jsbeautifier/javascript/beautifier.py [hunk 17]
```diff
@@ -848,18 +869,19 @@ def handle_word(self, current_token):
...
-            else:
-                if self._options.brace_style in ['expand', 'end-expand'] or (
-                        self._options.brace_style == 'none' and current_token.newlines):
-                    prefix = 'NEWLINE'
-                else:
-                    prefix = 'SPACE'
-                    self._output.space_before_token = True
+        is_literal_sbrk_cma = self._flags.mode == MODE.ObjectLiteral and self._last_last_text in ['{', ',']
+        token_reserved_word = reserved_array(current_token, ['else', 'catch', 'finally', 'from'])
+        no_brace_style_nl = self._options.brace_style == 'none' and current_token.newlines
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:874]: UNNECESSARY

File: python/jsbeautifier/javascript/beautifier.py [hunk 18]
```diff
@@ -848,18 +869,19 @@ def handle_word(self, current_token):
...
+        brace_style_in_exp = self._options.brace_style in ['expand', 'end-expand', 'expand-all']
+        if self._flags.last_token.type == TOKEN.END_BLOCK and self._previous_flags.inline_frame:
+            prefix = 'SPACE'
+        elif self._flags.last_token.type == TOKEN.END_BLOCK and not token_reserved_word:
+            prefix = 'NEWLINE'
+        elif self._flags.last_token.type == TOKEN.END_BLOCK and brace_style_in_exp or no_brace_style_nl:
+            prefix = 'NEWLINE'
+        elif self._flags.last_token.type == TOKEN.END_BLOCK:
+            prefix = 'SPACE'
+            self._output.space_before_token = True
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:884]: UNNECESSARY

File: js/src/javascript/beautifier.js [hunk 19]
```diff
@@ -543,6 +541,8 @@ Beautifier.prototype.handle_start_expr = function(current_token) {
         if (!this._options.keep_array_indentation) {
           this.print_newline();
         }
+      } else if (this._options.brace_style === "expand-all" && this._flags.last_token.text === ',') {
+        this.print_newline();
```
Comment (UAK-35) [inline js/src/javascript/beautifier.js:545]: ### _**UNNECESSARY**_

File: python/jsbeautifier/javascript/beautifier.py [hunk 20]
```diff
@@ -477,6 +486,10 @@ def handle_start_expr(self, current_token):
                     # }, [ goes to a new line
                     if not self._options.keep_array_indentation:
                         self.print_newline()
+                else:
+                    if self._options.brace_style == "expand-all" and \
+                            self._flags.last_token.text == ',':
+                        self.print_newline()
```
Comment (UAK-35) [inline python/jsbeautifier/javascript/beautifier.py:492]: ### _**UNNECESSARY**_

File: js/src/javascript/beautifier.js [hunk 21]
```diff
@@ -638,38 +641,30 @@ Beautifier.prototype.handle_start_expr = function(current_token) {
 Beautifier.prototype.handle_end_expr = function(current_token) {
   // statements inside expressions are not valid syntax, but...
   // statements must all be closed when their container closes
-  while (this._flags.mode === MODE.Statement) {
-    this.restore_mode();
-  }
-
+  while (this._flags.mode === MODE.Statement) { this.restore_mode(); }
```
[the remark this answers is not in the document]
Comment (bitwiseman) [reply js/src/javascript/beautifier.js:644]: ? Could you explain this comment?

**Comment by bitwiseman** [comment]
@UAK-35 Any update?

**Comment by bitwiseman** [comment]
@UAK-35  Any update?

File: .codeclimate.yml [hunk 22]
```diff
@@ -3,6 +3,17 @@ engines:
...
+  complex-logic:
+    enabled: false
+  file-lines:
+    enabled: false
+  method-lines:
+    enabled: false
+  nested-control-flow:
+    enabled: false
+  similar-code:
+    enabled: false
```
[the remark this answers is not in the document]
Comment (bitwiseman) [reply .codeclimate.yml:16]:
```suggestion
```

**Comment by bitwiseman** [comment]
Interesting idea.  Hopefully someone will pick it up some day.

**Comment by bitwiseman** [comment]
Interesting feature.  Hopefully someone will come and finish it some time.

**Pull request closed without merging**